src/hash.c

```diff
@@ -70,10 +70,13 @@ static uint32_t int_hash_scalar(const int* x) {
 static uint32_t dbl_hash_scalar(const double* x) {
   double val = *x;
   // Hash all NAs and NaNs to same value (i.e. ignoring significand)
-  if (R_IsNA(val)) {
-    val = NA_REAL;
-  } else if (R_IsNaN(val)) {
-    val = R_NaN;
+  // Avoid calling both expensive `R_IsNA()` and `R_IsNaN()` checks
+  if (isnan(val)) {
+    if (R_IsNA(val)) {
```

Comment:
We could presumably make this even faster with our own implementation to avoid testing `isnan()` twice: https://github.com/wch/r-source/blob/1a74f78a3c9e751f465b6eb27f7c1a8bf5ff2ada/src/main/arithmetic.c#L122-L126

Reply:
Or maybe we should have something that returns an enum so we could put in a switch statement instead?

Reply:
If you're interested in NAs vs NaNs there's a little more info at https://github.com/tidyverse/haven/blob/master/src/tagged_na.c#L7-L24

Reply:
I like the enum idea! 

- Check if `isnan()`, if not, return `vctrs_non_missing` (or some equivalent)
- If true, use the 1954 trick to either return `vctrs_na` or `vctrs_nan`

Reply:
Would also be useful in `vec_compare`